File: aula_emg/emg_labiocom.m
function [ rmsmed,integral,freqmedian,psdpeak,linenveli,rmsi,freqmediani,signal, temp ] = emg_labiocom(dat,freq,graficos,windowlength,overlap)
% emg_analisys
% Realiza a análise de EMG de 1 músculos
% Author: Prof. Dr. Paulo Roberto Pereira Santiago - EEFERP/USP
% 04/11/2013
%%
if nargin==1
    freq = 2000;
    windowlength = 3;
    overlap = 1.5;
    graficos = 1;
end
if nargin==2
    graficos = 1;
    windowlength = 0.125;
    overlap = 0.0625;
end
if nargin==3
    windowlength = 0.125;
    overlap = windowlength/2;
end

if nargin==4
    overlap = 0.0625;
end

windowlength = windowlength * freq;
overlap = overlap * freq;

% datemgvoltraw = importemg_delsys(dat);
datemgvoltraw = dat;
datemgvoltfilt = filtbutter(datemgvoltraw,[15 500],freq,'bandpass');
%

datemg = datemgvoltfilt(:,2)*1000000; % passando de volts para micro volts
% datemg = datemgvoltfilt(:,2);

% obl_ext_L = datemg(:,2);

%% tempo
[nl1,nc1] = size(datemg);
temp1 = [(0:nl1-1)/freq]'; % vetor tempo
%%
close all

%% GRAFICO 1 - RAW EMG

    figure(1)
%     subplot(2,2,1)
    plot(temp1,dat)
    % plot(rec_abd_L)
    ylabel ('Amplitude (V)');
    title('RAW EMG')
    axis tight
    grid on

%%

disp('Deseja selecionar uma intervalo de tempo para as análises?')
disp('Para "NÃO" digite 0')
disp('Para "SIM" digite quanto deseja retirar do início e fim em seg. i.e. [10 20]')
disp('para uma análise do 10 ao 20 segundo')

tanalise = input('Digite sua opção:  ');

if size(tanalise,2) == 1 && tanalise(1,1) == 0
    tini = 1;
    tfim = nl1;
end
if size(tanalise,2) == 2 && tanalise(1,1) == 0
        tini = 1;
        tfim = tanalise(1,2)*freq;
end

if size(tanalise,2) == 2 && tanalise(1,1) ~= 0
    tini = tanalise(1,1)*freq;
    tfim = tanalise(1,2)*freq;
end

signal = datemg(tini:tfim,:);

%% tempo
[nl,nc] = size(signal);
temp = [(0:nl-1)/freq]'; % vetor tempo
%%

%% GRAFICO 1 - RAW EMG
if graficos == 1
    figure(1)
    clf

    % subplot(2,2,1)
    plot(temp,signal)
    hold on
    % plot(rec_abd_L)
    ylabel ('Amplitude (µV)');
    title('EMG SELECT AND FILTER')
    grid on
    axis tight
end

%% SIGNAL RECTIFIED
signalD = detrend(signal);
signalDabs = abs(signalD);

%% GRAFICO 2 - FULL-WAVE RECTIFIED EMG
if graficos == 1
    figure(2)
    % subplot(2,2,2)
    plot(temp,signalDabs)
    hold on
    % plot(rec_abd_R_abs)
    ylabel ('Amplitude (µV)');
    title('FULL-WAVE EMG')
    axis tight
end

%% LINEAR ENVELOPE
signalenv = abs(filtbutter(signalDabs,50,freq,'low'));
signal_integ = trapz(temp,signalenv(:,2));



%% GRAFICO 3 - LINEAR ENVELOPE
if graficos == 1
    figure(2)
    % subplot(2,2,1)
    plot(temp,signalenv(:,2),'r','linewidth',2)
    % plot(rec_abd_L_linenvel(:,2))
    ylabel ('Amplitude (µV)');
    % title(['LEFT RECTUS ABDOMINIS (LINEAR ENVELOPE EMG = ',num2str(rec_abd_L_linenvel_integ),')'])
    title(['FULL-WAVE & LINEAR ENVELOPE = ',num2str(roundn(signal_integ,-1)),' µV.s)'])
    axis tight
end
%% RMS do sinal - window length e overlap
signalrms2 = rms2(signal,windowlength,overlap,'NO')'; 


%% 
length_rms2 = length(signalrms2);
temp_rms = [linspace(0,temp(end),length_rms2)]'; % vetor tempo rms
%%
%% RMS de todo o sinal
signalrms = rms(signal);

%% GRAFICO 4 - RMS do EMG
if graficos == 1
figure(3)
% subplot(2,2,1)
plot(temp_rms,signalrms2)
% plot(rec_abd_L_rms2)
hold on
line([temp_rms(1) temp_rms(end)],[signalrms , signalrms],'LineWidth',2,'LineStyle','--','Color','r')
% line([1 , length_rms2],[rec_abd_L_rms , rec_abd_L_rms],'LineWidth',2,'Line','--','Color','r')
ylabel ('RMS (µV)');
title(['EMG RMS = ',num2str(roundn(signalrms,-1)),' µV'])
xlabel('time (s)')
axis tight
end

%% Analise no domínio da frequência

[MPF_signal,PEAK_signal,F50_signal,F95_signal,F_signal,P_signal] = psd2(signal,freq);




%% GRAFICO 5 - PSD
if graficos == 1
    figure(4)
    % subplot(2,2,2)
    plot(F_signal,P_signal)
    ylabel ('Power (dB)');
    title('EMG - PSD')
    xlabel('Frequency (Hz)')
    axis tight
end

%% Frequencia mediana no dominio no temporal

Fs = freq;
t = temp;
nfft = 2^nextpow2(windowlength);
window = hanning(round(windowlength));
overlap = round(overlap);

% [signal_S,signal_F,signal_T,signal_P] = spectrogram(signal,window,overlap,window,Fs);
[signal_S,signal_F,signal_T,signal_P] = spectrogram(signal,window,overlap,nfft,Fs);
% [signal_P1,signal_F1] = pwelch(signal,window,overlap,nfft,Fs);

  

  %%
  medianfreqs_signal = ones(size(signal_P,2),1);

  
  %%
  
 for nn = 1:size(signal_P,2)
       % 
        signal_P_normcumsumpsd = cumsum(signal_P(:,nn))./sum(signal_P(:,nn));
        Ind1 = find(signal_P_normcumsumpsd <=0.5,1,'last');
        sizeInd1 = size(Ind1,1);
        
%         psdest = psd(spectrum.periodogram,signal,'Fs',2000,'NFFT',length(signal)); 
%    normcumsumpsd = cumsum(psdest.Data)./sum(psdest.Data);
%    Ind = find(normcumsumpsd <=0.5,1,'last');
%    fprintf('Median frequency is %2.3f Hz\n',psdest.Frequencies(Ind));
   
        
        if sizeInd1 == 0
            medianfreqs_signal(nn) = 3;
        elseif sizeInd1 ~= 0
            medianfreqs_signal(nn) = signal_F(Ind1);
        end
        
end

%    zerosfm1 = find(medianfreqs_signal == 0);
%    medianfreqs_signal(zerosfm1) = F50_signal; 
%    medianfreqs_signal1 = filtbutter(medianfreqs_signal,freq/10,freq,'low');
%   medianfreqs_signal1 = filtbutter(medianfreqs_signal,freq/10,freq,'low');
  medianfreqs_signal1 = smooth(medianfreqs_signal,0.5);
% medianfreqs_signalf = medianfreqs_signalf;
   medianfreqs_signalf = medianfreqs_signal1;
   freqmediani = medianfreqs_signal;
       
   tam_mf = length(medianfreqs_signalf);
   t_mft = [linspace(0,temp(end),tam_mf)]'; % vetor tempo freq median t
   %% GRAFICOS 6 - Frequencia Mediana na série temporal
 if graficos == 1
      figure(6)
    %   subplot(2,2,1)
      plot(t_mft,medianfreqs_signal);
      hold on
      plot(t_mft,medianfreqs_signalf,'color','r','linewidth',1);
      xlabel('Time (s)'); 
      ylabel('Hz');
      title('Median Frequency')
      axis tight
 end
%%
integral = signal_integ;
rmsmed = signalrms;
freqmedian = F50_signal;
psdpeak = PEAK_signal;
linenveli = signalenv;
rmsi = signalrms2;

end

%%

function [datf] = filtbutter(dat,fc,freq,ftype)
if nargin == 2
    freq = 100;
    ftype = 'low'; 
end

n=4; %ordem do filtro

wn=fc/(freq/2);   %frequencia de corte de

[b,a] = butter(n,wn,ftype); %definindo os parametros para o filtro de Butterworth

if size(dat,2) == 1
    dat = [[1:size(dat,1)]',dat];
end

[nlin,ncol] = size(dat);
 
datf = NaN(nlin,ncol);
for i = 2:ncol
datf(:,i) = filtfilt(b,a,dat(:,i));
end

datf = [[1:size(dat,1)]',datf(:,2:end)];

end
%%

function [y] = rms2(signal, windowlength, overlap, zeropad)
%% DECLARATIONS AND INITIALIZATIONS
% Calculates windowed (over- and non-overlapping) RMS of a signal using the specified windowlength
% y = rms(signal, windowlength, overlap, zeropad)
% signal is a 1-D vector
% windowlength is an integer length of the RMS window in samples
% overlap is the number of samples to overlap adjacent windows (enter 0 to use non-overlapping windows)
% zeropad is a flag for zero padding the end of your data...(0 for NO, 1 for YES)
% ex. y=rms(mysignal, 30, 10, 1).  Calculate RMS with window of length 30 samples, overlapped by 10 samples each, and zeropad the last window if necessary
% ex. y=rms(mysignal, 30, 0, 0).  Calculate RMS with window of length 30 samples, no overlapping samples, and do not zeropad the last window
%
% Author: A. Bolu Ajiboye


delta = windowlength - overlap;

%% CALCULATE RMS

indices = 1:delta:length(signal);
% Zeropad signal
if length(signal) - indices(end) + 1 < windowlength
    if zeropad
        signal(end+1:indices(end)+windowlength-1) = 0;
    else
        indices = indices(1:find(indices+windowlength-1 <= length(signal), 1, 'last'));
    end
end

y = zeros(1, length(indices));
% Square the samples
signal = signal.^2;

index = 0;
for i = indices
	index = index+1;
	% Average and take the square root of each window
	y(index) = sqrt(mean(signal(i:i+windowlength-1)));
end
end
%%
function varargout = psd2(varargin)
%PSD2 Power Spectral Density and frequency characteristics.
%  PSD2 estimates the power spectral density (PSD), mean power frequency (MPF),
%  peak frequency (PEAK), and limit frequency (F95) that contains up to 95%
%  of the PSD using Welch's averaged periodogram method (PSD Matlab function).
%  [MPF,PEAK,F50,F95,F,P]=psd2(X,Fs,NFFT,WINDOW,NOVERLAP,DFLAG)
%  [MPF,PEAK,F50,F95,F,P]=psd2(X,Fs) uses default values
%  Inputs:
%    X: data vector
%    Fs: sampling frequency
%    NFFT,WINDOW,NOVERLAP,DFLAG: see PSD matlab function for help
%      default values: NFFT=1024, WINDOW=256, NOVERLAP= WINDOW/2, DFLAG='mean'
%  Outputs:
%    MPF: mean power frequency
%    PEAK: peak frequency (mode)
%    F50: median frequency of the power spectral density
%    F95: frequency limit that contains up to 95% of the power spectral density
%    F: frequency vector	
%    P: PSD estimated vector
%
%   See also PSD
%  Marcos Duarte  mduarte@usp.br 11oct1998 

if nargin == 2
    x=varargin{1};
    fs=varargin{2};
    if length(x)<1000
        nfft=512;
        window=256;
    else
        nfft=1024;
        window=512;
    end
    noverlap=window/2;
    dflag='mean';
elseif nargin==6
    x=varargin{1};
    fs=varargin{2};
    nfft=varargin{3};
    window=varargin{4};
    noverlap=varargin{5};
    dflag=varargin{6};
else
    error('Incorrect number of inputs')
    return
end
%power spectral density:
% [p,f]=psd(x,nfft,fs,window,noverlap,dflag); % OBSOLETE in Matlab 2018
[p,f]=pwelch(x, hanning(nfft), noverlap, 1024, fs);
%mpf:
mpf=trapz(f,f.*p)/trapz(f,p);
%peak:
[m,peak]=max(p);
peak=f(peak);
%50 and 95% of PSD:
area=cumtrapz(f,p);
f50=find(area >= .50*area(end));
if ~isempty(f50)
    f50=f(f50(1));
else
    f50=0;
end
f95=find(area >= .95*area(end));
if ~isempty(f95)
    f95=f(f95(1));
else
    f95=0;
end
varargout{1}=mpf;
varargout{2}=peak;
varargout{3}=f50;
varargout{4}=f95;
varargout{5}=f;
varargout{6}=p;
end

File: aula_emg/emg_labiocom.py
# -*- coding: utf-8 -*-
"""
EMG Signal Processing
Author: Prof. PAULO ROBERTO PEREIRA SANTIAGO
Institution: University of São Paulo
             School of Physical Education and Sport of Ribeirão Preto
             Biomechanics and Motor Control Laboratory
Date: 07/03/2023

This Python script was created to perform electromyographic (EMG) signal processing.
Initially, the EMG signal file is loaded into the program, followed by an analysis that is
performed between two points in time, defined by the user. Subsequently, the signal is filtered
with a 15-450 Hz band-pass filter. After this step, an RMS (Root Mean Square) calculation is
performed using 1 second windows and 1/2 second overlap. The script also calculates the
median frequency for each window of the filtered EMG signal. The processing results in four
plots: raw EMG signal, filtered EMG signal, RMS EMG, and median frequency over time.

How to use:
The script is run via the command line and accepts four arguments:
python3 emg_labiocom.py [emg_file] [fs] [start_time] [end_time] [--no-plot]

Where:
emg_file: is the path to the EMG signal file you wish to process.
fs: is the sampling rate of the EMG signal in Hz.
start_time: is the start time for the signal analysis in seconds.
end_time: is the end time for the signal analysis in seconds.
--no-plot: is an option that, if specified, disables the plotting of the EMG signal plots.

Example:
python3 emg_labiocom.py emg_data.txt 2000 2 7 --no-plot

Output:
The script will produce four plots representing:
1. Raw EMG signal.
2. Filtered EMG signal.
3. RMS EMG.
4. Median frequency over time.
If the --no-plot option is specified, the plots will not be displayed.
Moreover, a text file with the calculated EMG results will be saved as {filename}_results_emg_labiocom.txt.

Prerequisites:
This script requires Python 3 and the following Python libraries: numpy, scipy, matplotlib, os, and argparse.
These can be installed via pip:
pip install numpy scipy matplotlib os argparse

Note:
Proper use of this script requires a solid understanding of signal processing concepts,
particularly as they pertain to EMG signal processing. Improper use can result in inaccurate
or misleading analyses. Therefore, it is recommended that users have adequate knowledge or
seek guidance from a specialist when using this script for EMG signal analysis.
"""

import numpy as np
import os
import argparse
import matplotlib.pyplot as plt
from scipy.signal import butter, filtfilt, welch, detrend
from scipy.signal.windows import hann


def butter_lowpass(cutoff, fs, order=4):
    nyq = 0.5 * fs
    normal_cutoff = cutoff / nyq
    b, a = butter(order, normal_cutoff, btype='low', analog=False)
    return b, a


def butter_lowpass_filter(data, cutoff, fs, order=4):
    b, a = butter_lowpass(cutoff, fs, order=order)
    y = filtfilt(b, a, data)
    return y


# Funcoes para o filtro de bandpass
def butter_bandpass(lowcut, highcut, fs, order=4):
    nyq = 0.5 * fs
    low = lowcut / nyq
    high = highcut / nyq
    b, a = butter(order, [low, high], btype='band')
    return b, a


def butter_bandpass_filter(data, lowcut, highcut, fs, order=4):
    b, a = butter_bandpass(lowcut, highcut, fs, order=order)
    y = filtfilt(b, a, data)
    return y


# Funcoes para retificar
def full_wave_rectification(emg_signal):
    emg_detrend = detrend(emg_signal)
    emg_abs = np.abs(emg_detrend)
    return emg_abs


# Funcoes para envelope linear
def linear_envelope(emg_abs, cutoff, fs):
    emg_envelope = butter_lowpass_filter(emg_abs, cutoff, fs)
    time = np.arange(len(emg_abs)) / fs
    signal_integ = np.trapz(emg_envelope, time)
    return emg_envelope, signal_integ


# Funcao para calculo do RMS
def calculate_rms(semg, window_length, overlap):
    start = 0
    rms_values = []

    while start + window_length < len(semg):
        window = semg[start:start+window_length]
        rms_value = np.sqrt(np.mean(window**2))
        rms_values.append(rms_value)
        start += overlap
    return rms_values


# Funcoes para calculo da frequencia mediana
def calculate_median_frequency(semg, fs, window_length, overlap):
    start = 0
    median_freq_values = []

    while start + window_length < len(semg):
        window = semg[start:start+window_length]
        median_freq_value = calculate_median_frequency_for_window(window, fs)
        median_freq_values.append(median_freq_value)
        start += overlap

    return median_freq_values


def calculate_median_frequency_for_window(window, fs):
    nperseg = len(window)
    noverlap = int(nperseg / 2)  # Escolha do overlap
    nfft = 1024  # Escolha do numero de pontos FFT
    freqs, psd = welch(window, fs, window=hann(nperseg), nperseg=nperseg, noverlap=noverlap, nfft=nfft)
    median_freq_idx = np.where(np.cumsum(psd) >= np.sum(psd) / 2)

    if median_freq_idx[0].size > 0:
        median_freq = freqs[median_freq_idx[0][0]]
    else:
        median_freq = np.nan  # ou algum outro valor que indique que a frequencia mediana nao pode ser calculada

    return median_freq


#  Funcao para calculo da curva de grau 2 para verificar tendencias nos resultados
def polynomial_fit(x, y, poly_deg=2):
    poly_coeff = np.polyfit(x, y, poly_deg)
    poly_vals = np.polyval(poly_coeff, x)

    return poly_vals


if __name__ == '__main__':
    # Argumentos da linha de comando
    parser = argparse.ArgumentParser(description='Processa um sinal EMG.')
    parser.add_argument('emg_file', help='O arquivo de sinal EMG para processar.')
    parser.add_argument('fs', type=int, help='A taxa de amostragem do sinal EMG em Hz.')
    parser.add_argument('start_time', type=float, nargs='?', default=0.0, help='O tempo de início para a análise do sinal em segundos.')
    parser.add_argument('end_time', type=float, nargs='?', default=None, help='O tempo final para a análise do sinal em segundos.')
    parser.add_argument('--no-plot', action='store_true', help='Se especificado, não plota os gráficos do sinal EMG.')
    args = parser.parse_args()

    # Carregar sinal
    emg_file = args.emg_file
    fs = args.fs

    # Passando de volts para micro volts µV.s
    emg_signal = np.loadtxt(emg_file)
    emg_signal[:, 1] = emg_signal[:, 1] * 1000000
    time_full = np.arange(len(emg_signal)) / fs

    # Definir início e fim da análise do sinal
    start_time = args.start_time

    # se end_time for None, calcular o fim do sinal com base no seu tamanho
    if args.end_time is None:
        end_time = len(emg_signal) / fs
    else:
        end_time = args.end_time

    start_index = int(start_time * fs)
    end_index = int(end_time * fs)
    emg_signal_cut = emg_signal[start_index:end_index, 1]

    # Filtrar o sinal bandpass
    lowcut = 10.0  # low frequency
    highcut = 450.0  # high frequency
    window_time = 0.25

    emg_filtered = butter_bandpass_filter(emg_signal_cut, lowcut, highcut, fs, order=4)

    # Retificacao completa da onda
    emg_detrend = detrend(emg_filtered)
    emg_abs = np.abs(emg_detrend)
    emg_envelope, signal_integ = linear_envelope(emg_abs, cutoff=10, fs=fs)

    # Calcular RMS
    window_length = int(fs * window_time)  # para uma janela de 250ms
    overlap = int(window_length / 2)  # para uma sobreposição de 125ms
    rms_values = calculate_rms(emg_filtered, window_length, overlap)

    # Calcular a frequencia mediana (MDF)
    window_length = int(fs * window_time)  # para uma janela de window_time ms
    overlap = int(window_length / 2)  # para uma sobreposição de metade window_time ms
    median_freq_values = calculate_median_frequency(emg_filtered, fs, window_length, overlap)

    # Criar vetores de tempo
    time = np.arange(start_index, end_index) / fs
    time_rms = np.arange(start_index, start_index+len(rms_values)*overlap, overlap) / fs

    # Ajustar polinômio de grau 2 ao RMS e plotar
    poly2_rms = polynomial_fit(time_rms, rms_values, 2)

    # Ajustar polinômio de grau 2 ao MDF e plotar
    poly2_mdf = polynomial_fit(time_rms, median_freq_values, 2)

    # Salvando o arquivo de resultado
    # Adicione o sufixo "_results_labiocom" ao nome do arquivo
    base = os.path.basename(args.emg_file)
    filename = os.path.splitext(base)[0]
    output_file = f"{filename}_results_emg_labiocom.txt"

    # Certifique-se de que rms_values e median_freq_values têm o mesmo tamanho
    assert len(rms_values) == len(median_freq_values), \
        "rms_values and median_freq_values must have the same length"

    # Converta as listas em arrays numpy para facilitar a manipulação
    time_array = np.array(time_rms)
    rms_array = np.array(rms_values)
    freq_array = np.array(median_freq_values)

    # Combine os arrays em uma única matriz 2D
    data_matrix = np.vstack((time_array, rms_array, freq_array)).T

    # Gravar dados no arquivo
    with open(output_file, 'w') as f:
        f.write(f"Linear envelope (µVolts.s): {signal_integ}\n")
        f.write("Time(s),RMS(µVolts),Median_Frequency(Hz)\n")
        np.savetxt(f, data_matrix, fmt='%f', delimiter=',')  # Agora a matriz será salva com separação por vírgulas

    print(f"Results written to {output_file}\n Have good studies!")

    if not args.no_plot:
        # Criar Graficos
        # Grafico do sinal bruto e filtrado
        # Cria os subplots
        fig, axs = plt.subplots(2, 1, figsize=(12, 8))

        # Grafico do sinal bruto
        axs[0].plot(time_full, emg_signal, label='Raw EMG', color='blue', linewidth=1)
        axs[0].set_title('Raw EMG Full-Signal')
        axs[0].set_xlabel('Time (s)')
        axs[0].set_ylabel('sEMG (µ Volts)')
        axs[0].axis('tight')
        axs[0].grid(True)

        # Grafico do sinal filtrado
        axs[1].plot(time, emg_signal_cut, label='Raw EMG', color='blue', linewidth=3)
        axs[1].plot(time, emg_filtered, label='Filtered EMG', color='red', linewidth=1)
        axs[1].set_title('Cut and Filtered EMG Signal')
        axs[1].set_xlabel('Time (s)')
        axs[1].set_ylabel('sEMG (µ Volts)')
        axs[1].axis('tight')
        axs[1].legend()
        axs[1].grid(True)

        # Ajusta o layout
        plt.tight_layout()

        # Grafico para Full-Wave Rectified EMG
        plt.figure(2)
        plt.plot(time, emg_abs)
        # Plotando o envelope linear
        plt.plot(time[:len(emg_envelope)], emg_envelope, color='r', linewidth=1)
        plt.xlabel('Time (s)')
        plt.ylabel('Rectified EMG (µ Volts)')
        plt.title(f'FULL-WAVE & LINEAR ENVELOPE = {signal_integ:.1f} µVolts.s')
        plt.grid(True)
        plt.axis('tight')

        # Plotando o RMS
        plt.figure(3)
        plt.plot(time_rms, rms_values)
        plt.plot(time_rms, poly2_rms, color='red', linestyle='--')
        plt.title('EMG - RMS')
        plt.xlabel('Time (s)')
        plt.ylabel('RMS (µ Volts)')
        plt.grid(True)

        # Plotando o MDF
        plt.figure(4)
        plt.plot(time_rms, median_freq_values)
        plt.plot(time_rms, poly2_mdf, color='red', linestyle='--')
        plt.title('EMG - Median Frequency')
        plt.xlabel('Time (s)')
        plt.ylabel('Frequency (Hz)')
        plt.axis('tight')
        plt.grid(True)

        # Plotando o MDF
        plt.figure(5)
        freqs, psd = welch(emg_filtered, fs)

        # Find the index of the maximum y value
        index_max = np.argmax(psd)

        # Find the corresponding x value
        freq_max = freqs[index_max]

        # Plotting the MDF
        plt.plot(freqs, psd)
        plt.title('EMG - PWelch')
        plt.ylabel('PSD (dB/Hz)')
        plt.xlabel('Frequency (Hz)')
        plt.axis('tight')
        plt.grid(True)

        # Add a red dot for the max
        plt.plot(freq_max, psd[index_max], 'ro')

        # Annotate the max value with an arrow
        plt.annotate('Max: {:.2f}, {:.2f}'.format(freq_max, psd[index_max]),
                    xy=(freq_max, psd[index_max]), xycoords='data',
                    xytext=(+10, +30), textcoords='offset points', fontsize=12,
                    arrowprops=dict(arrowstyle="->", connectionstyle="arc3,rad=.2"))

        plt.show()
